### ISSUE TITLE
[REGRESSION][`pytest-xdist == 3.6.0`] Makes pre-existing pytest configurations crash with `execnet.gateway_base.RemoteError: concurrent remote_exec would cause deadlock for main_thread_only execmodel`

### ISSUE THREAD

**Issue by webknjaz** [pytest-dev/pytest-xdist#1071]
`pytest-xdist == 3.6.0` caused a regression, it seems. See how it fails in setuptools' CI: https://github.com/pypa/setuptools/actions/runs/8759251151/job/24041851385?pr=4092#step:9:114. I was constantly pushing new changes to a PR, working on a CI fix for something else, and one minute it was fine but later on it started failing due to unrelated changes. When I checked the logs and verified that it's unpinned, I confirmed that the working CI was using `pytest-xdist==3.5.0` (like the last commit on `main`) while the broken one started using `pytest-xdist==3.6.0`.

Here's how it explodes:
```console
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/_pytest/main.py", line 273, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>                          ^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/_pytest/main.py", line 327, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pluggy/_hooks.py", line 501, in __call__
INTERNALERROR>     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pluggy/_manager.py", line 119, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pluggy/_callers.py", line 181, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pluggy/_result.py", line 99, in get_result
INTERNALERROR>     raise exc.with_traceback(exc.__traceback__)
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pluggy/_callers.py", line 166, in _multicall
INTERNALERROR>     teardown.throw(outcome._exception)
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/_pytest/logging.py", line 796, in pytest_runtestloop
INTERNALERROR>     return (yield)  # Run all the tests.
INTERNALERROR>             ^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pluggy/_callers.py", line 102, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/xdist/dsession.py", line 138, in pytest_runtestloop
INTERNALERROR>     self.loop_once()
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/xdist/dsession.py", line 163, in loop_once
INTERNALERROR>     call(**kwargs)
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/xdist/dsession.py", line 201, in worker_workerfinished
INTERNALERROR>     self.config.hook.pytest_testnodedown(node=node, error=None)
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pluggy/_hooks.py", line 501, in __call__
INTERNALERROR>     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pluggy/_manager.py", line 119, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pluggy/_callers.py", line 138, in _multicall
INTERNALERROR>     raise exception.with_traceback(exception.__traceback__)
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pluggy/_callers.py", line 102, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pytest_cov/plugin.py", line 316, in pytest_testnodedown
INTERNALERROR>     self.cov_controller.testnodedown(node, error)
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/pytest_cov/engine.py", line 342, in testnodedown
INTERNALERROR>     rinfo = node.gateway._rinfo()
INTERNALERROR>             ^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/execnet/gateway.py", line 87, in _rinfo
INTERNALERROR>     self._cache_rinfo = RInfo(ch.receive())
INTERNALERROR>                               ^^^^^^^^^^^^
INTERNALERROR>   File "/home/runner/work/setuptools/setuptools/.tox/py/lib/python3.11/site-packages/execnet/gateway_base.py", line 934, in receive
INTERNALERROR>     raise self._getremoteerror() or EOFError()
INTERNALERROR> execnet.gateway_base.RemoteError: concurrent remote_exec would cause deadlock for main_thread_only execmodel
```

_Originally posted by @webknjaz in https://github.com/pytest-dev/pytest-xdist/issues/1070#issuecomment-2067290245_

P.S. I haven't had a chance to investigate deeper. Just making an issue where other people can land and add context in the future.

**Comment by webknjaz** [comment]
@jaraco is it possible that https://github.com/jaraco/pytest-enabler/blob/adb94869501cfb83329962beafb0807f152aba2a/pytest_enabler/__init__.py#L101 interferes with the updates to `pytest-xdist` and `execnet` somehow?

**Comment by webknjaz** [comment]
Here's one backref with a successful pin update in a project for stats: https://github.com/ixm-one/pytest-cmake-presets/pull/254.

**Comment by zmedico** [comment]
I hope https://github.com/pytest-dev/execnet/pull/274 will serve as a simple fix.

**Comment by jaraco** [comment]
> @jaraco is it possible that https://github.com/jaraco/pytest-enabler/blob/adb94869501cfb83329962beafb0807f152aba2a/pytest_enabler/__init__.py#L101 interferes with the updates to `pytest-xdist` and `execnet` somehow?

Possible, but unlikely. If pytest-enabler is installed, you can disable its functionality with `-p no:enabler` and see if that has any effect.

**Comment by zmedico** [comment]
> I hope [pytest-dev/execnet#274](https://github.com/pytest-dev/execnet/pull/274) will serve as a simple fix.

I've also opened https://github.com/pytest-dev/pytest-xdist/pull/1072 with a one-liner workaround in pytest-xdist, in case it's more convenient to fix this without an execnet release.

**Comment by RonnyPfannschmidt** [comment]
https://pypi.org/project/pytest-xdist/3.6.0/ is yanked

**Comment by webknjaz** [comment]
@RonnyPfannschmidt could you also yank `execnet`? The previous version of `pytest-xdist` doesn't have a cap so it'll pull it in too: https://github.com/pytest-dev/pytest-xdist/blob/v3.5.0/setup.cfg#L47.

**Comment by webknjaz** [comment]
> could you also yank execnet?

Or even instead? Since with all `execnet` v2 yanked, the depresolvers would exclude `pytest-xdist == 3.6.0` automatically..

**Comment by RonnyPfannschmidt** [comment]
The execmodel isn't the default 
So a execnet yank is unnecessary

**Comment by webknjaz** [comment]
Ah, okay, good to know.

**Comment by zmedico** [comment]
I don't have a minimal reproducer for this issue, but it is reproduced by running `tox` in a setuptools checkout, and if either https://github.com/pytest-dev/execnet/pull/274 or https://github.com/pytest-dev/pytest-xdist/pull/1072 is applied to the .tox/py virtualenv then it fixes the issue and pytest-cov produces coverage results.

**Comment by zmedico** [comment]
Based on feedback I've updated https://github.com/pytest-dev/execnet/pull/274 to only add a test case, and updated comments in https://github.com/pytest-dev/pytest-xdist/pull/1072 to reflect that pytest-xdist needs to cache rinfo for main_thread_only backward compatibility with pytest-cov.

**Comment by bluetech** [comment]
Marking as fixed by #1072.
